### PR TITLE
bpo-38870: Use subTest in test_unparse for better error reporting

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue38870

### CODE REVIEW

File: Lib/test/test_unparse.py [hunk 1]
```diff
@@ -120,26 +120,30 @@ def assertASTEqual(self, ast1, ast2):
...
-        self.assertASTEqual(ast1, ast2)
+        with self.subTest(code1=code1, ast_parse_kwargs=kwargs):
+            ast1 = ast.parse(code1, **kwargs)
+            code2 = ast.unparse(ast1)
+            ast2 = ast.parse(code2, **kwargs)
+            self.assertASTEqual(ast1, ast2)
 
     def check_invalid(self, node, raises=ValueError):
-        self.assertRaises(raises, ast.unparse, node)
+        with self.subTest(node=node, raises=raises):
```
Comment (isidentical) [inline Lib/test/test_unparse.py]: do we need to keep this `raises` here? The error message already includes which error didn't raised

File: Lib/test/test_unparse.py [hunk 2]
```diff
@@ -120,26 +120,30 @@ def assertASTEqual(self, ast1, ast2):
...
+        with self.subTest(code1=code1, code2=code2):
+            code1, code2 = self.get_source(code1, code2)
+            self.assertEqual(code2, code1)
 
     def check_src_dont_roundtrip(self, code1, code2=None):
-        code1, code2 = self.get_source(code1, code2)
-        self.assertNotEqual(code2, code1)
+        with self.subTest(code1=code1, code2=code2):
+            code1, code2 = self.get_source(code1, code2)
+            self.assertNotEqual(code2, code1)
```
Comment (isidentical) [inline Lib/test/test_unparse.py]: In general the parameter `code2`  will be `None`, so I think we should use the code2 that comes from the get_source
```py
code1, code2 = get_source
with self.subTest(code1=code1, code2=code2): 
     run tests
```

File: Lib/test/test_unparse.py (hunk 1 above)
Comment (pablogsal) [reply Lib/test/test_unparse.py]: Meh, it does not hurt but I agree is redundant. I will remove it :+1:

File: Lib/test/test_unparse.py [hunk 3]
```diff
@@ -120,26 +120,30 @@ def assertASTEqual(self, ast1, ast2):
...
         code2 = code2 or code1
         code1 = ast.unparse(ast.parse(code1))
         return code1, code2
 
     def check_src_roundtrip(self, code1, code2=None):
-        code1, code2 = self.get_source(code1, code2)
-        self.assertEqual(code2, code1)
+        with self.subTest(code1=code1, code2=code2):
+            code1, code2 = self.get_source(code1, code2)
+            self.assertEqual(code2, code1)
```
Comment (isidentical) [inline Lib/test/test_unparse.py]:
```suggestion
        code1, code2 = self.get_source(code1, code2)
        with self.subTest(code1=code1, code2=code2):
            self.assertEqual(code2, code1)
```

**Review by isidentical — APPROVED** [review]
LGTM (since you marked `raises=` as resolved)

File: Lib/test/test_unparse.py [hunk 4]
```diff
@@ -120,13 +120,15 @@ def assertASTEqual(self, ast1, ast2):
...
-        self.assertASTEqual(ast1, ast2)
+        with self.subTest(code1=code1, ast_parse_kwargs=kwargs):
+            ast1 = ast.parse(code1, **kwargs)
+            code2 = ast.unparse(ast1)
+            ast2 = ast.parse(code2, **kwargs)
+            self.assertASTEqual(ast1, ast2)
 
     def check_invalid(self, node, raises=ValueError):
-        self.assertRaises(raises, ast.unparse, node)
+        with self.subTest(node=node, raises=raises):
```
Comment (pablogsal) [inline Lib/test/test_unparse.py]:
```suggestion
        with self.subTest(node=node):
```

**Comment by pablogsal** [comment]
> (since you marked `raises=` as resolved)

That was an error, it should have been removed :S

**Comment by isidentical** [comment]
> That was an error, it should have been removed :S

Ah, I see. Thanks

**Pull request merged**